[Lam Kai Yi Karen] iP

Karen-Lam opened the pull request:
# DukePro

> "Your mind is for having ideas, not holding them." - [David Allen](https://dansilvestre.com/productivity-quotes/)

### DukePro frees your mind of having to remember things you need to do. It's 
- text-based
- easy to learn
- ~~FAST~~  *SUPER*  FAST to use

### All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

### And it is **FREE**!

### Features:
 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

##
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
add Duke.java

Commit:
add Duke.java

Commit:
add Task.java
add Duke.java

Commit:
add Task.java
add Duke.java

Commit:
no message

Commit:
add Duke.java

Commit:
add Duke.java
add Event.java
add Task.java

Commit:
add Deadline.java
add Todo.java

Commit:
add Duke.java

Commit:
add EXPECTED.txt
add input.txt
add Deadline.class
add Duke.class
add Event.class
add Task.class
add Todo.class

Commit:
add Duke.java

Commit:
add Duke.java

Commit:
add Duke.java
add Deadline.java
add Todo.java
add Event.java
add Task.java
add duke.txt

Commit:
add Duke.java

Commit:
add duke.txt
add Deadline.java
add Duke.java
add Event.java
add input.txt

Commit:
Merge branch 'branch-Level-7'

Commit:
add duke.txt
add Duke.java
add AddDeadlineCommand.java
add AddEventCommand.java
add AddTodoCommand.java
add Command.java
add DeleteCommand.java
add ExitCommand.java
add ListCommand.java
add MarkCommand.java
add Parser.java
add Storage.java
add TaskList.java
add Ui.java
add UnmarkCommnd.java

Commit:
add AddDeadlineCommand.java
add AddEventCommand.java
add AddTodoCommand.java
add Command.java
add Deadline.java
add Delete.java
add DeleteCommand.java
add Duke.java
add Event.java
add ExitCommand.java
add ListCommand.java
add MarkCommand.java
add Parser.java
add Storage.java
add Task.java
add TaskList.java
add Todo.java
add Ui.java
add UnmarkCommand.java

Commit:
git config pull.rebase false  # merge

Commit:
add build.gradle
add gradlew
add gradlew.bat

Commit:
add build.gradle
add ParserTest.java
add TodoTest.java

Commit:
added gradle

Commit:
Added javadocs

Commit:
Revert "added gradle"

This reverts commit ef1c1b9964f9cbd1df55dc3ac729ca3a401d1cfa.

Commit:
Edit files to follow Coding Standards

Commit:
Add Find function

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Task.java

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/AddDeadlineCommand.java
#	src/main/java/Command.java
#	src/main/java/ExitCommand.java
#	src/main/java/ListCommand.java
#	src/main/java/Parser.java

Commit:
Edit to follow checkstyle

Commit:
Add GUI

russelltankaimin commented on src/main/java/Deadline.java:
```diff
@@ -6,9 +6,14 @@ public Deadline(String description, String by) {
         this.by = by;
     }
 
+    public Deadline(String description, boolean done, String by) {
```
Perhaps consider renaming `done` to `isDone` as per the coding style. Boolean field names should start with a verb.

russelltankaimin commented on src/main/java/Deadline.java:
```diff
@@ -6,9 +6,14 @@ public Deadline(String description, String by) {
     @Override
     public String toString() {
-        return "[D]" + super.toString() + " (by: " + by + ")";
+        return "D | " + super.toString() + " | by: " + by;
```
Perhaps consider using `String.format()` instead of just concatenating strings. It may be much concise. For your consideration 😊

russelltankaimin commented on src/main/java/Duke.java:
```diff
@@ -1,11 +1,15 @@
 
 public class Duke {
+    static String STR = "------------------------------------------------------------";
     static ArrayList<Task> arr = new ArrayList<Task>();
```
Perhaps a more fitting name instead of `arr` will be `tasks` or `taskLists` with emphasis on the plurality since it has a container for tasks.👍

russelltankaimin commented on src/main/java/Duke.java:
```diff
@@ -15,19 +19,59 @@ public static void main(String[] args) {
+            Scanner scannerOne = new Scanner(f);
+            while(scannerOne.hasNextLine()) {
+                String[] parts = scannerOne.nextLine().split(Pattern.quote(" | "));
+                switch (parts[0]) {
```
`switch` and `case` keywords should be indented on the same column

russelltankaimin commented on src/main/java/Duke.java:
```diff
@@ -15,19 +19,59 @@ public static void main(String[] args) {
+            while(scannerOne.hasNextLine()) {
+                String[] parts = scannerOne.nextLine().split(Pattern.quote(" | "));
+                switch (parts[0]) {
+                    case "T":
```
For tasks, maybe consider using Factory methods inside a parser. You can then encapsulate the parsing logic inside the parser. With that said, this may result in 
```java
case "D":
case "T":
case "E":
   parser.add(response);
   break;
```
Just a suggestion to keep code concise, but currently still looks good 👍👍. For your consideration.

Commit:
Edit to address comments

WesleyBLDC commented on src/main/java/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,56 @@
+package duke.command;
```
the .java files all seem to be in one folder, recommend to put related files into its own folder for easier management and viewing.

WesleyBLDC commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+     *
+     * @return String The numbered List of Tasks.
+     */
+    public String toStringList() {
```
Would this be redundant? With relation to the toString function that you already have.

WesleyBLDC commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public String toString() {
+        if (isDone) {
+            return "1" + " | " + this.name;
```
It is not very clear what the 1 and 0 is supposed to represent. Perhaps there is a more intuitive way to show it or to update the javadocs.

WesleyBLDC reviewed (approved):
Overall the code and logic looks good.

The only thing is that the .java and classes are all in one folder. I think there could be better file management.

Commit:
Edit Parser.java

Edit program to enable user to key case insensitive instructions keyword.

Previous instruction keyword input by user is case-sensitive. User has to key all letters in instruction in lower-case for instruction to be detected by program.

Add method to convert instruction by user to lowercase.

Commit:
Add GUI

Edit program to program a Graphic User Interface to make program more
user friendly.

Previously, user relies solely on Command Line Interface to interact
with program.

Let's:
- Add DialogBox.fxml and MainWindow.fxml for formatting GUI.
- Edit methods in Duke.java so that String outputs from Parser.java is
directed to GUI.

Commit:
Merge branch 'branch-Level-10'

Commit:
Edit Parser.java to use Varargs

Commit:
Merge branch 'branch-A-Varargs'

Commit:
Edit files to follow Checkstyle

Commit:
Merge branch 'branch-A-Checkstyle'

Commit:
Edit Parser.java to include Assertions

Parser: include assert statements to check for correct assumptions

Previously, Parser did not include assert statements.

Let's,
- create assert statement to check that number of word input by user
corresponds to the instruction indicated by user
- create assert statement to check that date keyed in by user is in
correct format of DD MMM YYYY HH:MM
- create assert statement to check that number keyed in for selection
of task to mark/delete/unmark can be parsed to integer

Commit:
Edit files to improve Code Quality

Storage loadData method: extract addTask method

Storage loadData method has issues of arrowhead code and too long
method.

Extracting the parsing of a line from storage to addTask method allows
code to be more readable.

Commit:
Edit TaskList.java to include Streams

TaskList findRelated method: Use Streams to implement method.

Previously, findRelated method was implemented by for loop, making it
look complicated and less readable.

Update findRelated method to use Streams, improving readability.

Karen-Lam committed:
Merge pull request #2 from Karen-Lam/branch-A-Assertions

Edit Parser.java to include Assertions

Commit:
Merge branch 'master' of https://github.com/Karen-Lam/ip

* 'master' of https://github.com/Karen-Lam/ip:
  Edit Parser.java to include Assertions

Karen-Lam committed:
Merge pull request #3 from Karen-Lam/branch-A-Streams

Edit TaskList.java to include Streams

Karen-Lam committed:
Merge branch 'master' into branch-A-CodeQuality

Karen-Lam committed:
Merge pull request #4 from Karen-Lam/branch-A-CodeQuality

Edit files to improve Code Quality

Commit:
Merge branch 'master' of https://github.com/Karen-Lam/ip

* 'master' of https://github.com/Karen-Lam/ip:
  Edit TaskList.java to include Streams
  Edit files to improve Code Quality

Commit:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Add A-CI

Commit:
Edit to include Update Command

Update Command: Allows user to edit previously inputted data, without
having to delete and add new task

Let's,
- create new updateCommand that extends from Command class to show how
to execute command
- add more conditional statements in parser to handle update command

Commit:
Merge branch 'branch-BCD-Extension'

# Conflicts:
#	src/main/java/Interface/Parser.java

Commit:
Update GUI

Edit to allow for resizing of window.

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Edit User Guide

Commit:
Merge branch 'branch-A-UserGuide'

Commit:
Updated GUI

Commit:
Merge branch 'branch-A-Personality'

Commit:
Edit User Guide

Commit:
Add more test case

Commit:
Edit to suit checkstyle

Fix update command bug

Commit:
Merge branch 'branch-A-MoreTesting'

# Conflicts:
#	src/test/java/ParserTest.java

Commit:
no message

Commit:
Add Ui image

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Change Ui image

Commit:
Change Ui image

Commit:
Edit user guide

Commit:
Fix storage issue

Commit:
Edit User Guide
